![logo-gt.png](attachment:logo-gt.png)

<div class="alert alert-block alert-info">
    <h1>BIOS 4150/BIOL 6150</h1>
    <h3>Instructor: Dr. King Jordan</h3>
    <p>TAs: Shivam Sharma (shivamsharma13@gatech.edu) & Vishva Natarajan (vishvanata@gatech.edu)</p>
</div>

<div class="alert alert-block alert-warning">
    <h2>Project 2 (Read Mapping) starter notebook</h2>
    <h3>Deadline: 11:59PM, October 31st, 2023</h3>
</div>

---

# **1. Getting ready for read alignment**
### *Total Questions: 5*
### *Total Points: 5+20+20+20+5 = 70*


---

<div class="alert alert-block alert-warning">
    <h3>1.1 Do you have the post-QCed fastq file from the assigned 1000 Genomes individual</h3>
    <p>We will start from where we left off in Project #1. We want to make sure that we are using the correct QCed fastq files.</p>
</div>

In [3]:
##Run "ls -lh" here to show the two fastq files.
!ls -lh ~/scratch/project/trimming/SRR*

-rw-r--r-- 1 akrishna311 gtperson 1.4G Oct 17 17:28 /home/hice1/akrishna311/scratch/project/trimming/SRR070501_1_trimmed.fastq.gz
-rw-r--r-- 1 akrishna311 gtperson 1.5G Oct 17 17:28 /home/hice1/akrishna311/scratch/project/trimming/SRR070501_2_trimmed.fastq.gz


<div class="alert alert-block alert-warning">
    <h3>1.2 Download the correct reference genome for alignment step</h3>
    <p>Please locate and donwload the human reference genome fasta file in build GRCh38</p>
</div>

In [4]:
#Where does this file come from? Give the website link (FTP?) or explain how you downloaed this file.

# We downloaded the file from FTP site of NCBI - ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.40_GRCh38.p14/
!wget ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.40_GRCh38.p14/GCF_000001405.40_GRCh38.p14_genomic.fna.gz


--2023-10-27 15:21:04--  ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000/001/405/GCF_000001405.40_GRCh38.p14/GCF_000001405.40_GRCh38.p14_genomic.fna.gz
           => ‘GCF_000001405.40_GRCh38.p14_genomic.fna.gz’
Resolving ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)... 130.14.250.13, 130.14.250.12, 2607:f220:41e:250::13, ...
Connecting to ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)|130.14.250.13|:21... connected.
Logging in as anonymous ... Logged in!
==> SYST ... done.    ==> PWD ... done.
==> TYPE I ... done.  ==> CWD (1) /genomes/all/GCF/000/001/405/GCF_000001405.40_GRCh38.p14 ... done.
==> SIZE GCF_000001405.40_GRCh38.p14_genomic.fna.gz ... 972898531
==> PASV ... done.    ==> RETR GCF_000001405.40_GRCh38.p14_genomic.fna.gz ... done.
Length: 972898531 (928M) (unauthoritative)

GCF_000001405.40_GR 100%[===================>] 927.83M  35.5MB/s    in 26s     

2023-10-27 15:21:31 (35.6 MB/s) - ‘GCF_000001405.40_GRCh38.p14_genomic.fna.gz’ saved [972898531]



In [7]:
#Also run "ls -lh" here to see the size and location of your reference file.
!ls -lh 
!mkdir ~/scratch/project/
!mv GCF_000001405.40_GRCh38.p14_genomic.fna.gz ~/scratch/project/

total 932M
-rw-r--r-- 1 akrishna311 gtperson 928M Oct 27 15:21 GCF_000001405.40_GRCh38.p14_genomic.fna.gz
-rw-r--r-- 1 akrishna311 gtperson  15K Oct 27 15:23 Project2-ReadMapping.ipynb


In [8]:
!ls -lh ~/scratch/project/

total 4.7G
drwxr-xr-x 2 akrishna311 gtperson 4.0K Oct 17 17:37 fastqc_after_trimming
drwxr-xr-x 2 akrishna311 gtperson 4.0K Oct 17 17:16 fastqc_before_trimming
-rw-r--r-- 1 akrishna311 gtperson 928M Oct 27 15:21 GCF_000001405.40_GRCh38.p14_genomic.fna.gz
-rw-r--r-- 1 akrishna311 gtperson 1.9G Oct 16 12:59 SRR070501_1.filt.fastq.gz
-rw-r--r-- 1 akrishna311 gtperson 2.0G Oct 16 13:00 SRR070501_2.filt.fastq.gz
drwxr-xr-x 2 akrishna311 gtperson 4.0K Oct 17 17:33 trimming


In [9]:
#MD5 checksum to ensure the complete download of refernece genome file.
#Checksum provided on NCBI website - c30471567037b2b2389d43c908c653e1
!md5sum ~/scratch/project/GCF_000001405.40_GRCh38.p14_genomic.fna.gz

c30471567037b2b2389d43c908c653e1  /home/hice1/akrishna311/scratch/project/GCF_000001405.40_GRCh38.p14_genomic.fna.gz


<div class="alert alert-block alert-warning">
    <h3>1.3 Find out about your reference genome file.</h3>
    <p>Does your reference fasta file have more information that just ACGT nucleotides? Please check if your fastq sequence have Ns or small/upper case ACGT nnucleotides.</p>
</div>

In [13]:
#Just glance over the fasta file. There is no need to run a script for this. 
#Write in one sentence if you see more than just upper case ACGT nucleotides.
!zcat ~/scratch/project/GCF_000001405.40_GRCh38.p14_genomic.fna.gz | head -200

# we have N's, a's, c's, g's, t's along with A,C,G,T bases

>NC_000001.11 Homo sapiens chromosome 1, GRCh38.p14 Primary Assembly
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN

<div class="alert alert-block alert-warning">
    <h3>1.4 Using 1-3 lines, explain what upper case ACGT, lower case ACGT, and NNNs mean in your reference fasta file.</h3>
    <p>Hint: check readme files.</p>
</div>

In [8]:
#No need to exceed 3 lines for this question.
# ACGT: Uppercase letters in a DNA sequence represent the unambiguous and known nucleotides in the sequence. These are the four standard DNA bases.
# acgt: Lowercase letters in a DNA sequence may indicate a lower confidence in the base call or may represent a mixed base, where the actual base cannot be determined with certainty.
# NNNs: The letter 'N' represents an ambiguous or unknown nucleotide. It is used when the specific base at a particular position cannot be determined, or when the sequence contains a mixture of bases.

<div class="alert alert-block alert-warning">
    <h3>1.5 Select a tool for read alignment.</h3>
    <p>We have demonstrated bwa and minimap2 in the class. If you want to use some other tool, reach out to one of the TAs to get that tool installed on PACE-ICE.</p>
</div>

In [14]:
#Run your tool here to just make sure bash does not say "not found"
!minimap2

Usage: minimap2 [options] <target.fa>|<target.idx> [query.fa] [...]
Options:
  Indexing:
    -H           use homopolymer-compressed k-mer (preferrable for PacBio)
    -k INT       k-mer size (no larger than 28) [15]
    -w INT       minimizer window size [10]
    -I NUM       split index for every ~NUM input bases [8G]
    -d FILE      dump index to FILE []
  Mapping:
    -f FLOAT     filter out top FLOAT fraction of repetitive minimizers [0.0002]
    -g NUM       stop chain enlongation if there are no minimizers in INT-bp [5000]
    -G NUM       max intron length (effective with -xsplice; changing -r) [200k]
    -F NUM       max fragment length (effective with -xsr or in the fragment mode) [800]
    -r NUM[,NUM] chaining/alignment bandwidth and long-join bandwidth [500,20000]
    -n INT       minimal number of minimizers on a chain [3]
    -m INT       minimal chaining score (matching bases minus log gap penalty) [40]
    -X           skip self and dual mappings (for the all-vs-all m

# **2. Read alignment**
### *Total Questions: 4*
### *Total Points: 30+50+10+40 = 130*


---

<div class="alert alert-block alert-warning">
    <h3>2.1 Steps before mapping</h3>
    <p>You already have a fasta file, but sometimes alignment tools do not accept fasta files directly. Please read your aligners documentation and run all the preliminary steps require to run the mapping command from your aligner.</p>
</div>

<div class="alert alert-block alert-info">
    <p>Preliminary commands go in the cells below. Please use appropriate comments and markdown texts to explain what your command is doing.</p>
</div>

In [15]:
#Write your commands here. Insert more cells if necessary
!minimap2 -d ~/scratch/project/GCF_000001405.40_GRCh38.p14_genomic.mmi ~/scratch/project/GCF_000001405.40_GRCh38.p14_genomic.fna.gz

[M::mm_idx_gen::34.246*1.77] collected minimizers
[M::mm_idx_gen::43.319*2.02] sorted minimizers
[M::main::49.129*1.90] loaded/built the index for 705 target sequence(s)
[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 705
[M::mm_idx_stat::49.761*1.89] distinct minimizers: 100224366 (37.12% are singletons); average occurrences: 5.896; average spacing: 5.582; total length: 3298430636
[M::main] Version: 2.26-r1175
[M::main] CMD: minimap2 -d /home/hice1/akrishna311/scratch/project/GCF_000001405.40_GRCh38.p14_genomic.mmi /home/hice1/akrishna311/scratch/project/GCF_000001405.40_GRCh38.p14_genomic.fna.gz
[M::main] Real time: 49.929 sec; CPU: 94.183 sec; Peak RSS: 12.120 GB


> Minimap2 generates a minimizer index file(.mmi) for the reference before mapping which reduces time taken for indexing. This .mmi file will be used for alignment.

In [2]:
#Run "ls -lh" to show what additional files were created in addition to your original fasta files.
!ls -lh ~/scratch/project/GCF*

-rw-r--r-- 1 akrishna311 gtperson 928M Oct 27 15:21 /home/hice1/akrishna311/scratch/project/GCF_000001405.40_GRCh38.p14_genomic.fna.gz
-rw-r--r-- 1 akrishna311 gtperson 7.2G Oct 27 15:58 /home/hice1/akrishna311/scratch/project/GCF_000001405.40_GRCh38.p14_genomic.mmi


<div class="alert alert-block alert-warning">
    <h3>2.2 Run mapping.</h3>
    <p>This should be one or two lines of code but it's important to understand what the options are doing.</p>
</div>

In [ ]:
#Write you command below.

In [18]:
#Add one extra cell as required and explain each argument/flag that was given to the mapping command.
!minimap2 -c -a ~/scratch/project/GCF_000001405.40_GRCh38.p14_genomic.mmi ~/scratch/project/trimming/SRR070501_1_trimmed.fastq.gz ~/scratch/project/trimming/SRR070501_2_trimmed.fastq.gz > ~/scratch/project/SRR070501_Minimap.sam  


[M::main::5.500*0.88] loaded/built the index for 705 target sequence(s)
[M::mm_mapopt_update::6.738*0.90] mid_occ = 748
[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 705
[M::mm_idx_stat::7.364*0.91] distinct minimizers: 100224366 (37.12% are singletons); average occurrences: 5.896; average spacing: 5.582; total length: 3298430636
[M::worker_pipeline::90.122*2.83] mapped 6496217 sequences
[M::worker_pipeline::159.656*3.00] mapped 6496215 sequences
[M::worker_pipeline::230.621*3.04] mapped 6496279 sequences
[M::worker_pipeline::243.865*3.02] mapped 1690820 sequences
[M::worker_pipeline::326.331*3.02] mapped 6496217 sequences
[M::worker_pipeline::394.688*3.05] mapped 6496215 sequences
[M::worker_pipeline::465.052*3.06] mapped 6496279 sequences
[M::worker_pipeline::478.281*3.05] mapped 1690820 sequences
[M::main] Version: 2.26-r1175
[M::main] CMD: minimap2 -c -a /home/hice1/akrishna311/scratch/project/GCF_000001405.40_GRCh38.p14_genomic.mmi /home/hice1/akrishna311/scratch/proj

> #### Options: <br>
-c - output CIGAR <br>
-a - output in the SAM format <br>

<div class="alert alert-block alert-warning">
    <h3>2.3 Post alignment evaluation.</h3>
    <p>How many lines does your SAM file have?</p>
</div>

In [19]:
#Just one line for this one.
!wc -l ~/scratch/project/SRR070501_Minimap.sam

65506564 /home/hice1/akrishna311/scratch/project/SRR070501_Minimap.sam


<div class="alert alert-block alert-warning">
    <h3>2.4 Look at the 7th read that maps successfully in your SAM file and answer the following questions. (10+10+10+10 = 40Points)</h3>
    <p>1. What chromosome does it align to?</p>
    <p>2. What is the CIGAR string for this read? Explain the CIGAR string in detail.</p>
    <p>3. Does this read have a paired read which also aligns successfully?</p>
    <p>4. If the paired read aligns successfully, what does the CIGAR string for this read convey?</p>
</div>

In [ ]:
# Grep for all lines except header lines
!grep -v "^@" ~/scratch/project/SRR070501_Minimap.sam

SRR070501.133	0	NC_000003.12	84427354	43	77M	*	0	0	GAGATGAAGAATGTTTTCCCTGAAATTATTAGTTGGCTTAACACAGATGAGCAGAAAGTTTGTAAATGTGAAGACAT	F?EFBF5FFFFDE?BDEEFFC?EEBEFFBFBEDEBFEE=DDEEEAEE?EEEEED=EBB=BBE=A@@@B>BADB@@D@	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:P	cm:i:12	s1:i:65	s2:i:0	de:f:0	rl:i:0
SRR070501.135	16	NC_000019.10	8958403	0	77M	*	0	0	CTGGCATGAATGTAGGATCATAGTAGGTGACTTCTGCCCTGGAGACCTCAGTAGCAGCATCCAGGGACACACTGGAG	?AA@?@5BBB@@A@D<????A@BB@B?=B??B=B5@A@A=F?DE@FEEGF?EGGFBBC?AGADGFGEF:GGGGGGGG	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:P	cm:i:11	s1:i:64	s2:i:64	de:f:0	rl:i:0
SRR070501.135	272	NW_025791807.1	208171	0	77M	*	0	0	*	*	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:S	cm:i:11	s1:i:64	de:f:0	rl:i:0
SRR070501.139	0	NC_000003.12	121694655	45	76M1S	*	0	0	ATTTTTTATTTGGGCATCTGTAACATCCTGACAGTAATTCCCAATGCTTCCATTAAGTTCTGCTAATTGATTCATAC	GEFFFF?FDFFFFDFAFFDDFF?FFFDFDFDFFFFF:FFFFFFF:D?EEEDEFFD=CBBEEE=C?:EEE-E?BCBE5	NM:i:0	ms:i:152	AS:i:152	nn:i:0	tp:A:P	cm:i:10	s1:i:66	s2:i:0	de:f:0	rl:i:0
SRR070501.141	16	NC_000

SRR070501.744	16	NC_000002.12	99284624	55	77M	*	0	0	ATGTATTATATTTGTCCCTATTAATTCTGTGCTTGAGACAACTGACTGATAGCGTTACACGTACCTCTCAGGTACTG	#?===DDADBA@=A5EE@EDB:B5=BAEDEBCBB:BBBBB>CC@=?BBBB@BD-DC=C@CCC1EEFFAFFDADBBDG	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:P	cm:i:11	s1:i:72	s2:i:0	de:f:0	rl:i:0
SRR070501.745	16	NC_000019.10	43072321	60	77M	*	0	0	TGGCCTGGCCCACAGAGGAACAAAAGATACAGAGGACATTCAGGGTGACTGGGTCACTGTGGATGCCACCATATCGG	################@-6???@?@>>:0>=::?@=>==-B@C@C<;>::?5=BE@AAA?AB:CCCC=A:AEBBDEE	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:P	cm:i:11	s1:i:64	s2:i:42	de:f:0	rl:i:0
SRR070501.745	272	NC_000019.10	43194964	0	3S25M4D49M	*	0	0	*	*	NM:i:5	ms:i:133	AS:i:130	nn:i:0	tp:A:S	cm:i:5	s1:i:42	de:f:0.0267	rl:i:0
SRR070501.745	272	NC_000019.10	42868725	0	3S25M4D49M	*	0	0	*	*	NM:i:5	ms:i:133	AS:i:130	nn:i:0	tp:A:S	cm:i:5	s1:i:42	de:f:0.0267	rl:i:0
SRR070501.745	272	NC_000019.10	43175839	0	3S25M4D49M	*	0	0	*	*	NM:i:7	ms:i:121	AS:i:118	nn:i:0	tp:A:S	cm:i:5	s1:i:42	de:f:0.0533	rl:i:0
SRR070501.746	16	NC_000

SRR070501.1554	272	NC_000019.10	42732981	0	3S74M	*	0	0	*	*	NM:i:3	ms:i:130	AS:i:130	nn:i:0	tp:A:S	cm:i:6	s1:i:53	de:f:0.0405	rl:i:0
SRR070501.1555	16	NC_000011.10	60937595	46	1S76M	*	0	0	GCACCATGAGAAAGGTAACCCTTCAGCCAGAGATGGCAGAGTTCCAGAAGCAGAAAGGAGCCTTGGAGGTCCACTAA	'6/99;:?7:::???39.6;56-7)->===?B@=@BB@@@E@BD?BB:@B9BBCA?B@AC:??@C@?C>>:?@B=FA	NM:i:0	ms:i:152	AS:i:152	nn:i:0	tp:A:P	cm:i:9	s1:i:67	s2:i:0	de:f:0	rl:i:0
SRR070501.1557	16	NC_000011.10	18360673	0	77M	*	0	0	AAGAAAAGATTCGGAGACAGTATTTAAGCACAAATGTAAGGCAGCAATCTGATTTTTGCCTGATCTTCTTTCTCTTT	EEEEGEDEEE@FDEEEDEEGEGFGGGFGGGGGGGGGBGGGGGAEGD?GGGEGGGGGBGGFFEGGGGGGGEGGEGDGG	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:P	cm:i:11	s1:i:70	s2:i:70	de:f:0	rl:i:0
SRR070501.1557	272	NW_021160006.1	109499	0	77M	*	0	0	*	*	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:S	cm:i:11	s1:i:70	de:f:0	rl:i:0
SRR070501.1558	16	NC_000004.12	163472211	48	77M	*	0	0	TGTGACTTTATCATTGACACCGTGGCGGACCACCTTGGCCTGGCCAATCTCAAAATTTTCTTGTGGGGTATAAATAA	####################???B:;9;59:4BBB????

SRR070501.2305	0	NC_000022.11	30093897	48	77M	*	0	0	ATTAATTATTTTTTAATAGGTTGGACTTGTTGAAATAATCCTGATACATTCCTACAATGGCCACTGCTCAGCTTTCT	FCFDFEFGGEGGGGGGGGGGGFGGGGGFGGGDGGFGFBGGGGGFGG?GGGGEGGEGGGGGGFGGGGF?FFFFDFFGG	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:P	cm:i:12	s1:i:68	s2:i:0	de:f:0	rl:i:0
SRR070501.2307	0	NC_000009.12	128693588	44	77M	*	0	0	TTGGGTGTTATGGAGAGATTGTATCAAAAGCTCTTCTGGTATTCATTTAGGATTTGACGAAACGTTCGAGTCAAACG	GGDCDDCFFFFFGGFGDEEEEEGFGFFEGGFEGGGEGGGEED?EEEEGGDGG?FEFBFFFDFFFE=FECEB?BEEBE	NM:i:1	ms:i:148	AS:i:148	nn:i:0	tp:A:P	cm:i:8	s1:i:66	s2:i:0	de:f:0.0130	rl:i:0
SRR070501.2308	16	NC_000005.10	138329607	55	77M	*	0	0	TGCTAAGATTCGAAAGATCGAGGCAACGTTTTGGGGTTCCTCTGTTGAGACATAATAGTACATCGAAATTCCCATTA	ACEEEEEEE?EEEEEEEEEEEEEEEEGGEGEEEGEFGEGGGGDGGFFGGGGGGGGFGGGGGGEGGGGGDFGGGGGGG	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:P	cm:i:14	s1:i:72	s2:i:0	de:f:0	rl:i:0
SRR070501.2309	16	NC_000001.11	156851239	50	77M	*	0	0	GTTGGAGGAAGGAGTGTAATAGAAGGAGCTGGTCAGAGGCCTAACCCTTACCCATCCACCATGGCGTCTCCCCGGAT	<:(@@C?CA

SRR070501.3001	272	NT_187607.1	727708	0	77M	*	0	0	*	*	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:S	cm:i:10	s1:i:61	de:f:0	rl:i:0
SRR070501.3001	272	NC_000016.10	18328024	0	77M	*	0	0	*	*	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:S	cm:i:10	s1:i:61	de:f:0	rl:i:0
SRR070501.3001	272	NC_000016.10	18368168	0	77M	*	0	0	*	*	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:S	cm:i:10	s1:i:61	de:f:0	rl:i:0
SRR070501.3001	256	NC_000016.10	14941954	0	77M	*	0	0	*	*	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:S	cm:i:10	s1:i:61	de:f:0	rl:i:0
SRR070501.3001	256	NC_000016.10	16340448	0	77M	*	0	0	*	*	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:S	cm:i:10	s1:i:61	de:f:0	rl:i:0
SRR070501.3002	16	NC_000009.12	41189084	0	77M	*	0	0	GTGACTGGGATGGTTAATTACCTAAATACCCGGTGATAATTGTGACTGGGATCTTGGCGCCGAGGCCAGACTTTTCC	#########################?33??9129/3?:7?=A?B@@@@5?@?=DBBBABEEDEEEEFEDEDEEEBFF	NM:i:1	ms:i:148	AS:i:148	nn:i:0	tp:A:P	cm:i:10	s1:i:60	s2:i:60	de:f:0.0130	rl:i:0
SRR070501.3002	272	NC_000009.12	178795	0	77M	*	0	0	*	*	NM:i:1	ms:i:148	AS:i:148	nn:

SRR070501.3570	256	NC_000008.11	2079333	0	77M	*	0	0	*	*	NM:i:1	ms:i:148	AS:i:148	nn:i:0	tp:A:S	cm:i:6	s1:i:53	de:f:0.0130	rl:i:0
SRR070501.3572	0	NC_000012.12	122288408	38	77M	*	0	0	TTCTACTATAAGTTACCACACATATCATGTTCTGACATCTTATTAAACACACACACATACAACAATAAACAGAAAGC	DBCADDDDDDD5ADDBDD5DDDDDDDDDD=BDDDB:B-A@>9ABDD?D4CBDC:BDB5@5@>B??:BBBCB=-=>6@	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:P	cm:i:12	s1:i:67	s2:i:0	de:f:0	rl:i:15
SRR070501.3573	0	NC_000003.12	140657935	39	77M	*	0	0	AAAGAAACACTAAATAAGAAAGGAAAAACTGTTACCACCCACTACAAAAACAGGCTGAATTCCACAGACCAGTGACA	GGEFAGEGGEGGGEGFGGGGEGGGGGGFGECEGGGGFGFEDGEEEFEEEE=@4,A?ACCCCCEEE@ECCCCA=:ACA	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:P	cm:i:10	s1:i:63	s2:i:0	de:f:0	rl:i:0
SRR070501.3574	16	NC_000007.14	144378775	0	77M	*	0	0	ACAGTGCCTTCGAGCCTACAAGCCCCGAGAGAATGATGAATTGGCACTGGAGAAAGCCGACGTGGTGATGGTGACTC	6<=?<=>:=5)???@??9AA:BBD@@B@CBDEABDBBDFBB=EEE@BAEEEFEEEAFEFEFEDFFFFFEFFFFFDFE	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:P	cm:i:11	s1:i:70	s2:i:70	de:f:0	rl:i:0
SRR070501.3574

SRR070501.4376	0	NC_000001.11	46932706	38	77M	*	0	0	TTGACCAGAGACTTCCCTCATTCCTCTATTCGAATTACCACACAGGACGTCTCACCTTGATCCTCCTGAGAAGGGCA	GGGFGGGGBGEEGFFFGGGEGEDGGGGEDGGGBGEFGGGDGEFEGEEDEABDFDDC?C@AACEEEEAA=C<?CBA?<	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:P	cm:i:13	s1:i:62	s2:i:0	de:f:0	rl:i:0
SRR070501.4377	16	NC_000002.12	210600612	53	77M	*	0	0	CTTCAGTTGAGTCCATTATGGCTACGGAAGACAGGCAGCTGTTTTCAGATAAACTAAATGAGATCAATGAAAAGATT	EEEEEEE?EEEEEEEEEEEEFE?CDFF:AFDEEFGFDGBFGGGFGEEGFGGGGDGGGFBGGGGGEGFGGGGGFGAFF	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:P	cm:i:10	s1:i:71	s2:i:0	de:f:0	rl:i:0
SRR070501.4378	0	NC_000009.12	121302002	38	77M	*	0	0	GCAGGGAAGGAGCCTGGCCTGCAGATCTGGCGTGTGGAGAAGTTCGATCTGGTGCCCGTGCCCACCAACCTTTATGG	GGFDDGEBFF@FAFEDEGEEEEEE:BECFEDF@BACC:>6>>99?>=???=?EEAEE=;?:?###############	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:P	cm:i:11	s1:i:62	s2:i:0	de:f:0	rl:i:0
SRR070501.4379	0	NC_000002.12	218738419	39	77M	*	0	0	AGCCTGTCTCCTGGCATCATTCAGGGGGTACTGGAGACTGTGCACCGCAGCCTGTTGACCATAAGGTGCCCAAAAGC	GDGDEGGFFFFAFA

SRR070501.5012	0	NC_000016.10	28390184	0	77M	*	0	0	CTGGCTTTTCAAGCACTCTGACTTAGTATAATCCTCACAAAGCTCTAAACCTAGTAATATTCCCCACAATTTTCCTC	FFGFEGGF=DDBA?EBBGFFG=FE=FD?EF=?BBEBEEFBBA=BBEBE?E?=BB,CBBBB=BEEE?EEEEEA=A,=:	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:P	cm:i:10	s1:i:63	s2:i:63	de:f:0	rl:i:0
SRR070501.5012	272	NC_000016.10	28725043	0	77M	*	0	0	*	*	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:S	cm:i:10	s1:i:63	de:f:0	rl:i:0
SRR070501.5013	0	NC_000002.12	52747172	35	77M	*	0	0	AAGTTTGGGAATCTGCCTTCTGACCTTCAGGAGGCATATTGGGCAGAGATGAGCCTTACAACGTCTCTAGCCTTCCT	EGEEEEEEDEEEGGG?BGFFGGBDGGFFGFFFGGBEAFFFEEEGDDBGFFFEGBBFBEBFAEDADEEDBFBBBBD@B	NM:i:1	ms:i:148	AS:i:148	nn:i:0	tp:A:P	cm:i:8	s1:i:61	s2:i:0	de:f:0.0130	rl:i:0
SRR070501.5016	16	NC_000003.12	184323896	57	77M	*	0	0	CATTGACCTCATTTTTGAGAAGGCCATTTCAGAGCCCAACTTCTCTGTGGCCTATGCCAACATGTGCCGCTGCCTCA	EE@EEACEE??CBC=@BGEEDBGAGEEDDEGEEGGEFECFBDGEEGGDFGGGGGGGGBGFEGGGGGFGGGGGEGGGG	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:P	cm:i:10	s1:i:73	s2:i:0	de:f:0	rl:i:0
SRR070501.5017	0

SRR070501.5664	0	NC_000001.11	212080447	52	77M	*	0	0	CTGAAAGCAAGAAATGGATGTACATGATCTTTAAGATATTATTTTCTATATCCTGTTTATTAGACAAAATAATTATC	GGGDGGEGGGFDGGFGDEFGDGGGGGEFGGGGEGGFGDFGEGGGGEFGGEEEEFFEFFEGGEFBDBA7BBB<EEBEE	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:P	cm:i:11	s1:i:70	s2:i:0	de:f:0	rl:i:0
SRR070501.5665	0	NC_000005.10	100856197	45	77M	*	0	0	TCTCTCCTCCTTTGACCATGAAAGCAGGAATCCAAAGGACACTGTCATTCAGCATGGAAAGTCTATGCACAAATTTT	GGGGGGGGEFGGGFEGGBEFEFGEFEDBBEDEDD@DB:=:>1??????CD@C=AACC;:AA9:3/>>?A>A######	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:P	cm:i:10	s1:i:66	s2:i:0	de:f:0	rl:i:0
SRR070501.5666	0	NC_000023.11	13043053	38	77M	*	0	0	AGTGATGACATGATCCAAAGCTGACTTTTCCTAAAAGACCTTATCTGACAGGCAAAAGGGACTTAAACACTTTACCC	E?EDEEEEE?EADED?DDA=DDDBDEEDEBEEDEEEEE=EA?D?CDCBEAEEEB-BBB-@BBDD@:DBB@A=DD@@:	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:P	cm:i:10	s1:i:62	s2:i:0	de:f:0	rl:i:0
SRR070501.5667	16	NC_000006.12	152433814	59	77M	*	0	0	ATGTCCATGGCAGATGACGAAGTTTCCAAGGCATTGAGTTCTTTTTCTTTCTCAGTTATCCAGACGGACAGAGTCTC	&6616<4);2-:?=

SRR070501.6289	0	NC_000001.11	196429564	0	77M	*	0	0	TGGTTTTGTTTACCTGTATGGGTAGAACCACAAGAGCAACACAAATCATAGCAACTACAAAAAGCTTGGAGGACCAT	GGGGGGGGGGGGGGGFGGGGGFGGGFGGGGGEGGBGGGGGGFFFGGGFGEFFEGGGEGGEEEEEE:@DCCDCEEEEE	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:P	cm:i:10	s1:i:62	s2:i:62	de:f:0	rl:i:0
SRR070501.6289	256	NW_025791754.1	279240	0	77M	*	0	0	*	*	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:S	cm:i:10	s1:i:62	de:f:0	rl:i:0
SRR070501.6290	4	*	0	0	*	*	0	0	TCACACCTGTAATCCTAGCACTTTGGGAGGCCAAGGAGGGAGGATTGCTTGAAGCCAGGAGTTGGAGACCAGCCTGG	CGFFDFFFFFAEFFFBFFFFFF=D?BEEFBBFA:BA@AAA?ADA@DDDA?EEBDEDB?5??:?:???EEEEEEEE?E	rl:i:60
SRR070501.6291	16	NC_000011.10	59652898	54	77M	*	0	0	AGAGACCCAATCCTTTTCCCGCTGCAACATGAGATTGGCATATGGATCCTTTCGGAGATGATCTTGATGACTGCTCC	BAC=AB@B@=<BEBCEEGEGEGEGGEEFEFFAGDEFEFGGFGGGEGGFDFGGG?GGGGGGGFGGFGGGGFGGGGGGF	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:P	cm:i:10	s1:i:64	s2:i:47	de:f:0	rl:i:0
SRR070501.6291	272	NW_009646206.1	299331	0	77M	*	0	0	*	*	NM:i:3	ms:i:136	AS:i:136	nn:i:0	tp:A:S	cm:i:7	s1:i:

SRR070501.6883	16	NC_000018.10	57188724	57	77M	*	0	0	ATCTGTCTTTACTAAAAGATGCTACATGTTGATGTAAGGTCTGTGGGTTACACTCAGAAAGCAACCACAAGCAGACA	:?D??ABBAA?A7AABBBB?BBB@?@B=BCFEEEAEEEDBFFEDDEBEDCFEFFFE=EDBCDFEFEFAFFDFFDFFF	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:P	cm:i:9	s1:i:73	s2:i:0	de:f:0	rl:i:0
SRR070501.6884	16	NC_000001.11	149512830	0	77M	*	0	0	TCAATTTCATGTTTTCAACGAAGGTTGAATTACTCCTACTGACATGGCTGTTGGTTTTCATTGCAGTAGATGTTTAG	@E=EEEEEEEEEEEEEEEGFEEGGGEGFGFDEGGGGGEGGGEGGGGGFGGGGGGGGGGGGGGGGGGGGGGGGGGGGG	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:P	cm:i:10	s1:i:67	s2:i:67	de:f:0	rl:i:0
SRR070501.6884	272	NC_000001.11	149550797	0	77M	*	0	0	*	*	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:S	cm:i:10	s1:i:67	de:f:0	rl:i:0
SRR070501.6884	272	NC_000001.11	149546060	0	77M	*	0	0	*	*	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:S	cm:i:10	s1:i:67	de:f:0	rl:i:0
SRR070501.6884	272	NC_000001.11	149527092	0	77M	*	0	0	*	*	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:S	cm:i:10	s1:i:67	de:f:0	rl:i:0
SRR070501.6884	272	NC_000001.11	149531821	0	77M	*

SRR070501.7498	16	NC_000002.12	62222463	18	1S76M	*	0	0	GGGGGGGGGCGGGCGGGCTCTCCAATATAAGCCATCTGAACTACTGCGAACCTGACCTGAGGGTCACGTCGGTGGTT	###################?7<?@@1B<@@=@@?<?7????::??>=>=6????:AA3BACBB@=DAEEFGGGGGGG	NM:i:2	ms:i:140	AS:i:140	nn:i:0	tp:A:P	cm:i:6	s1:i:51	s2:i:0	de:f:0.0263	rl:i:0
SRR070501.7499	0	NC_000010.11	68691105	24	77M	*	0	0	TTTCACAGCTTGGCGAAGTGGCTCCTCTCCCCACCCTGTCTGCTCCTGTGATGGAGCCCCTCATTCATTCTTAGCCT	GGGGGFGGFGGGGGEEGEEDEEGEG:GADCAD?E?A7BBBBA=?@CA##############################	NM:i:2	ms:i:142	AS:i:142	nn:i:0	tp:A:P	cm:i:8	s1:i:55	s2:i:0	de:f:0.0260	rl:i:0
SRR070501.7500	0	NC_000003.12	52578993	41	77M	*	0	0	TTTCTACTGGTTCCAAATTTCTCAGATTCAACCTCATCTTCCCTAATTCAATGACACTCACCTGCCCACACGAGGAC	GFGGGGGGGGGGFGGGGEGFGGGFGGGGGGFGGGGGGGGGGFGGBDFEGFEEDADFECGFEDBEBCEDEGFEEE?EE	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:P	cm:i:10	s1:i:64	s2:i:0	de:f:0	rl:i:0
SRR070501.7501	16	NC_000009.12	127891025	55	77M	*	0	0	GGCACCGAGAGGGCAGTGTCTGGTGGATAAGGAAAGTCAACATTATCACGGCCACTCTGTGCTGGCCCTGTGCTCCA	#####

SRR070501.8117	256	NW_017852933.1	957347	0	77M	*	0	0	*	*	NM:i:1	ms:i:148	AS:i:148	nn:i:0	tp:A:S	cm:i:9	s1:i:57	de:f:0.0130	rl:i:0
SRR070501.8118	16	NT_187693.1	124443	0	77M	*	0	0	GGGGCGTGGCCCCAGGCTCAGGGAACAACTCAGGGGGACCCAGCCTCCTGGTGCCACTGCCTGTGAATCCTCCCAGC	#########?73?8::?=5?AAA79:;0;35=>>>D=BDDC?CDDEEEEECDEFF?EEGDGEGGGFFEGGEGGGGEG	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:P	cm:i:9	s1:i:62	s2:i:62	de:f:0	rl:i:0
SRR070501.8118	272	NW_003571054.1	124443	0	77M	*	0	0	*	*	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:S	cm:i:9	s1:i:62	de:f:0	rl:i:0
SRR070501.8118	272	NW_003571061.2	124443	0	77M	*	0	0	*	*	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:S	cm:i:9	s1:i:62	de:f:0	rl:i:0
SRR070501.8118	272	NW_003571058.2	124443	0	77M	*	0	0	*	*	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:S	cm:i:9	s1:i:62	de:f:0	rl:i:0
SRR070501.8118	272	NW_003571056.2	124443	0	77M	*	0	0	*	*	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:S	cm:i:9	s1:i:62	de:f:0	rl:i:0
SRR070501.8118	272	NC_000019.10	54149595	0	77M	*	0	0	*	*	NM:i:0	ms:i:154	AS:i:154	nn:i:0	

SRR070501.8755	0	NC_000010.11	92463938	50	77M	*	0	0	ACTGGATGATGAATCTCAAGCCCTGTATGCCATTAGCTCGACGTGGCCCGCTGAAGACGATATAGCCTGAAACACAG	DEEDE=EEA?EEEEBDEDDEEEEEEBEEEE:EEEBEEABC-@@=@A>D@:?+:55;6???????@=<:?=??A?A@@	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:P	cm:i:10	s1:i:69	s2:i:0	de:f:0	rl:i:0
SRR070501.8756	0	NC_000001.11	160634772	13	77M	*	0	0	CCGCCTTTTCACTCCAGCTGTAAGCCACATGGTCCCCCTTCTCCACTGTGCAGCCCAGCATCAAGGTGCAGGTCCCG	EEFFFFFGGFGGGGGGGGGGEDEGGEGEFGDGEDGGDEEEEEEEEEEDAECC?CAC:?,?;?;;?############	NM:i:1	ms:i:148	AS:i:148	nn:i:0	tp:A:P	cm:i:7	s1:i:48	s2:i:0	de:f:0.0130	rl:i:0
SRR070501.8757	0	NC_000007.14	133644916	36	77M	*	0	0	TGCAAACTATTTAAAGTTTTGTGAGCCACAGTGCTCAGCCATAGCTGTATCCCCTTTGCATCTTTGTATGTGTTTTC	FGGGGGGGGGGGFGFFGGGGGGGGGGFGGGGGGGFGGGGGGFGGGFGGFGGGGGGEEDEEEEEEGEECFECBCDDEC	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:P	cm:i:9	s1:i:61	s2:i:0	de:f:0	rl:i:0
SRR070501.8758	16	NC_000001.11	70266	11	77M	*	0	0	TAGTATTTGTGTACACCTGTTCCTTCCTGTGTGACCCTAAGTTTAGTAGAAGAAAGGAGAGAAAATATAGCCTAGCT	C:C?C;=@99@6;C.

IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.

Current values:
NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
NotebookApp.rate_limit_window=3.0 (secs)



In [29]:
# Question 2.4.3
# Grep for matching position ID to find paired read
!grep "20366027" ~/scratch/project/SRR070501_Minimap.sam

SRR070501.145	16	NC_000017.11	20366027	0	77M	*	0	0	AAAAGTAAGTATTAAGAAAGATAAAATATTTTTCAACCTTCCTGAAAGAAAATGTAAAGTAATATTTGGTTATGGCT	<9:=;7=?<:C=/A@D?DD<?5EDD@>5?EBBBEB??EBBEEEEACEDEBEEE?EAAA?ABA+A?:AAA5EEEE:AA	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:P	cm:i:15	s1:i:73	s2:i:73	de:f:0	rl:i:0
SRR070501.20366027	0	NC_000005.10	69116233	41	77M	*	0	0	TTTTTATTTTAACAAATTTCTATTTATGGATTTTGATGGTCACATCATTTACTTATTTTGGGAAATTCTTCAGTGCT	GFGGGFGGGGGEGFGGGGGAGAFFFFF?EBEEFFFFFFFFGGFEGG:GGGDDGGFGGFE:?DAEEEEGGGECGGFBD	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:P	cm:i:12	s1:i:64	s2:i:0	de:f:0	rl:i:0
SRR070501.20366027	16	NC_000005.10	69116412	48	77M	*	0	0	CTCCCTCTAAGAGAGGACAAAAAGGTACCCTTATTGGATATTCTCCTGAAGGAACACCTCTTTATAACTTCATGGGT	C@DEC:E:EEBDFEFEEBAGFGFEGF?EEEEEGGGFGDGEGFGEFGGGGGGGEFGEFGGEGGGGGG?GFEFGGGGGG	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:P	cm:i:10	s1:i:68	s2:i:0	de:f:0	rl:i:0


> #### 7th read that was successfully mapped - <br>
SRR070501.145	16	NC_000017.11	20366027	0	77M	*	0	0	AAAAGTAAGTATTAAGAAAGATAAAATATTTTTCAACCTTCCTGAAAGAAAATGTAAAGTAATATTTGGTTATGGCT	<9:=;7=?<:C=/A@D?DD<?5EDD@>5?EBBBEB??EBBEEEEACEDEBEEE?EAAA?ABA+A?:AAA5EEEE:AA	NM:i:0	ms:i:154	AS:i:154	nn:i:0	tp:A:P	cm:i:15	s1:i:73	s2:i:73	de:f:0	rl:i:0 <br>

> 1. The read aligns to chromosome 17(Searched for NC_000017.11 in NCBI website - https://www.ncbi.nlm.nih.gov/nuccore/568815581
> 2. CIGAR string - 77M which means there are 77 matches in the given read with respect to reference genome.
> 3. We did not find a paired read for this read which aligned successfully. We tried to grep for position and SRRID which did not return any sucessful matches.
> 4. No match

# **2. Post Read alignment**
### *Total Questions: 5*
### *Total Points: 10+20+25+15+30 = 100*


---

<div class="alert alert-block alert-warning">
    <h3>3.1 Samtools is a classic bioinformatics tools.</h3>
    <p>Check if you have samtools installed on your environment.</p>
</div>

In [30]:
#Just run the "samtools" command here and make sure it does not say "samtools not found".
!samtools


Program: samtools (Tools for alignments in the SAM format)
Version: 1.17 (using htslib 1.18)

Usage:   samtools <command> [options]

Commands:
  -- Indexing
     dict           create a sequence dictionary file
     faidx          index/extract FASTA
     fqidx          index/extract FASTQ
     index          index alignment

  -- Editing
     calmd          recalculate MD/NM tags and '=' bases
     fixmate        fix mate information
     reheader       replace BAM header
     targetcut      cut fosmid regions (for fosmid pool only)
     addreplacerg   adds or replaces RG tags
     markdup        mark duplicates
     ampliconclip   clip oligos from the end of reads

  -- File operations
     collate        shuffle and group alignments by name
     cat            concatenate BAMs
     consensus      produce a consensus Pileup/FASTA/FASTQ
     merge          merge sorted alignments
     mpileup        multi-way pileup
     sort           sort alignment file
     split          splits a

<div class="alert alert-block alert-warning">
    <h3>3.2 Is your SAM file sorted by coordinates?</h3>
    <p>SAM or BAM (which we will create later) can be sorted by the read name or read coordinate. Look at your SAM file and see if you can run a Samtools command to find out if your file is sorted. (Hint: Google your question)</p>
</div>

In [32]:
#Tell us if your SAM file is sorted.
#Using head command to check if file is sorted
!head ~/scratch/project/SRR070501_Minimap.sam

@HD	VN:1.6	SO:unsorted	GO:query
@SQ	SN:NC_000001.11	LN:248956422
@SQ	SN:NT_187361.1	LN:175055
@SQ	SN:NT_187362.1	LN:32032
@SQ	SN:NT_187363.1	LN:127682
@SQ	SN:NT_187364.1	LN:66860
@SQ	SN:NT_187365.1	LN:40176
@SQ	SN:NT_187366.1	LN:42210
@SQ	SN:NT_187367.1	LN:176043
@SQ	SN:NT_187368.1	LN:40745


In [33]:
#To check if file is sorted using samtools command. "SO:unsorted" tells us that this SAM file is unsorted.
!samtools view -H ~/scratch/project/SRR070501_Minimap.sam

@HD	VN:1.6	SO:unsorted	GO:query
@SQ	SN:NC_000001.11	LN:248956422
@SQ	SN:NT_187361.1	LN:175055
@SQ	SN:NT_187362.1	LN:32032
@SQ	SN:NT_187363.1	LN:127682
@SQ	SN:NT_187364.1	LN:66860
@SQ	SN:NT_187365.1	LN:40176
@SQ	SN:NT_187366.1	LN:42210
@SQ	SN:NT_187367.1	LN:176043
@SQ	SN:NT_187368.1	LN:40745
@SQ	SN:NT_187369.1	LN:41717
@SQ	SN:NC_000002.12	LN:242193529
@SQ	SN:NT_187370.1	LN:161471
@SQ	SN:NT_187371.1	LN:153799
@SQ	SN:NC_000003.12	LN:198295559
@SQ	SN:NT_167215.1	LN:155397
@SQ	SN:NC_000004.12	LN:190214555
@SQ	SN:NT_113793.3	LN:209709
@SQ	SN:NC_000005.10	LN:181538259
@SQ	SN:NT_113948.1	LN:92689
@SQ	SN:NC_000006.12	LN:170805979
@SQ	SN:NC_000007.14	LN:159345973
@SQ	SN:NC_000008.11	LN:145138636
@SQ	SN:NC_000009.12	LN:138394717
@SQ	SN:NT_187372.1	LN:40062
@SQ	SN:NT_187373.1	LN:38054
@SQ	SN:NT_187374.1	LN:176845
@SQ	SN:NT_187375.1	LN:39050
@SQ	SN:NC_000010.11	LN:133797422
@SQ	SN:NC_000011.10	LN:135086622
@SQ	SN:NC_000012.12	LN:133275309
@SQ	SN:NC_000013.11	LN:114364328
@SQ	SN:NC_000014.9	LN:10704

#Tell us how you reached your answer and clerly write the commands used in additional cell with descriptive comments.

> To determine if a SAM file is sorted, we can check the header lines in the SAM file.
The header should contain a line that indicates the sorting order of the alignments.
###### Format: @HD     VN:1.6  SO:coordinate(for genomic postion)\ SO:queryname(read name) SO:unsorted no SO field

<div class="alert alert-block alert-warning">
    <h3>3.3 Sort your SAM file?</h3>
    <p>Sort your SAM file by genomic coordinates,</p>
    <p>When you run this command, what is the output format of your result file?</p>
    <h4>Tweak samtools and change your sorted SAM file's output to SAM. (Hint: You should be able to read the file using simple cat or less)</h4>
</div>

In [36]:
#Run the sort command here.
# samtools will sort by genomic coordinates by default. 
# -o: output destination 
# -@: specifies the number threads
!samtools sort ~/scratch/project/SRR070501_Minimap.sam -o ~/scratch/project/SRR070501_Minimap.sam -O sam -@ 32

[bam_sort_core] merging from 0 files and 32 in-memory blocks...


In [37]:
#ls -lh to show sorted sam file was created, we overwrote the unsorted sam file with sorted sam file
!ls -lh ~/scratch/project/

total 26G
drwxr-xr-x 2 akrishna311 gtperson 4.0K Oct 17 17:37 fastqc_after_trimming
drwxr-xr-x 2 akrishna311 gtperson 4.0K Oct 17 17:16 fastqc_before_trimming
-rw-r--r-- 1 akrishna311 gtperson 928M Oct 27 15:21 GCF_000001405.40_GRCh38.p14_genomic.fna.gz
-rw-r--r-- 1 akrishna311 gtperson 7.2G Oct 27 15:58 GCF_000001405.40_GRCh38.p14_genomic.mmi
-rw-r--r-- 1 akrishna311 gtperson 1.9G Oct 16 12:59 SRR070501_1.filt.fastq.gz
-rw-r--r-- 1 akrishna311 gtperson 2.0G Oct 16 13:00 SRR070501_2.filt.fastq.gz
-rw-r--r-- 1 akrishna311 gtperson  15G Oct 27 19:06 SRR070501_Minimap.sam
drwxr-xr-x 2 akrishna311 gtperson 4.0K Oct 17 17:33 trimming


In [38]:
#Head your sorted SAM file here. Show the first 5 lines.
!head ~/scratch/project/SRR070501_Minimap.sam

@HD	VN:1.6	SO:coordinate
@SQ	SN:NC_000001.11	LN:248956422
@SQ	SN:NT_187361.1	LN:175055
@SQ	SN:NT_187362.1	LN:32032
@SQ	SN:NT_187363.1	LN:127682
@SQ	SN:NT_187364.1	LN:66860
@SQ	SN:NT_187365.1	LN:40176
@SQ	SN:NT_187366.1	LN:42210
@SQ	SN:NT_187367.1	LN:176043
@SQ	SN:NT_187368.1	LN:40745


<div class="alert alert-block alert-warning">
    <h3>3.4 Convert your SAM file to a BAM file.</h3>
    <p>This should be a single one line command.</p>
</div>

In [3]:
#Write the command run "ls -lh" to show the size and location of the file.
!ls -lh ~/scratch/project/SRR070501_Minimap.sam

-rw-r--r-- 1 akrishna311 gtperson 15G Oct 27 19:06 /home/hice1/akrishna311/scratch/project/SRR070501_Minimap.sam


In [40]:
#samtool command to convert SAM file to BAM file
# -b: convert to BAM format
# -o: output file
# -@: number of threads
!samtools view -b ~/scratch/project/SRR070501_Minimap.sam -o ~/scratch/project/SRR070501_Minimap.bam -@ 32

In [41]:
!ls -lh ~/scratch/project/SRR070501_Minimap.bam

-rw-r--r-- 1 akrishna311 gtperson 2.8G Oct 27 19:13 /home/hice1/akrishna311/scratch/project/SRR070501_Minimap.bam


<div class="alert alert-block alert-warning">
    <h3>3.5 Statistics of your SAM file.</h3>
    <p>1.How many total raw sequences does your SAM file have?</p>
    <p>2.How many total reads are mapped in your SAM file?</p>
</div>

In [43]:
#Show your commands and answers in additional cells as required.
!tail -10 ~/scratch/project/SRR070501_Minimap.sam

SRR070501.23904968	4	*	0	0	*	*	0	0	TACCTTCATATTTTATGTGCAGTGTATAATACACACTTTGGGTACTTTGATATTTTCTGTACAGTATGCTATATATT	FCDFEFC?A>CCEEBCDFFFDFBFEFADAECBEE?B?DBBA:@A=DACC5B:EEEEAC:-A@BB=B:CBB?B5>>>A	rl:i:0
SRR070501.23904977	4	*	0	0	*	*	0	0	TAGGCCAATTTCAAGTTCAGCAATGATTAACACTGGCAGAAGCTCCAGAGCGGGACCTGGTGGCCGCAGGGCTCGCG	B@:=5-?,;?<)@AAC5::5:-,648:B=AB:B?B==-5*99:==:>52:5>@>==@####################	rl:i:0
SRR070501.23904994	4	*	0	0	*	*	0	0	GAGGTCCAAGGATGGAAGCAAAGGATGGCAGCCAAGGGGCTTGCAAGGCAGAACATGGCCTTAGGCTTTAAGCTGCT	ED=5+>=ADDAADE?C5EE=EEE:-5E->A-A@..@>,>,:=:A5DC-=:5-6@94=A-B>?=@-5-:A:?-=5BD-	rl:i:0
SRR070501.23905013	4	*	0	0	*	*	0	0	ACTTTGGGAGGCCGAGGTGGGTGGATCACAAGGTCAGCAGATCAAGACCATCCTGGCCAACATGGTGCAACCCCCGT	B5DDDDD:B-?;?BA?:7>:5'>?=6@C5CC5?=CC5AAAA-:DD:DC?=-:C>CC5;@@9:?D=?=*?>AAAB?4=	rl:i:59
SRR070501.23905035	4	*	0	0	*	*	0	0	ACCAAGATTTTTGGCTTTCTGTACAAAGCTGTTTACACACAGGAAGGGGCAGGCATAACGGCTGCTAGGAGCTGTGG	-?A3?555A?:6A->A?==-@-ACA-;A--A-;;5;<-AC:>-CC--6-@.C=..66AA5AC:A:BB-.,47AA###	rl:i:0
SRR070501

In [44]:
#Number of raw reads in SAM file
!grep -v "^@" ~/scratch/project/SRR070501_Minimap.sam | wc -l 

65505857


In [46]:
#You can parse the SAM file manually, or figure out a faster way to do this.
# 4th column is set to 0 for unmapped reads without coordinate.
!grep -v "^@" ~/scratch/project/SRR070501_Minimap.sam | awk '{if ($4 > 0) {print $0}}' | wc -l

63883103


In [4]:
#Verifying total reads using samtool command
# -c: to get the count of reads
!samtools view -c ~/scratch/project/SRR070501_Minimap.sam

65505857


In [6]:
#Verifying mapped reads using samtool command
# -cF 4: excludes matches which are unmapped by checking 4th column(POS)
!samtools view -cF 4 ~/scratch/project/SRR070501_Minimap.sam

63883103
